examples/algorithms/bc/bc.cu

```diff
@@ -54,51 +53,63 @@ void test_bc(int num_arguments, char** argument_array) {
   // --
   // GPU Run
 
+  size_t n_runs = source_vect.size();
   std::vector<float> run_times;
-  for (int i = 0; i < source_vect.size(); i++) {
-    // Record run times without collecting metrics (due to overhead)
-    run_times.push_back(gunrock::bc::run(G, source_vect[i], true,
-                                         bc_values.data().get(), &edges_visited,
-                                         &search_depth));
-  }
-
-  // --
-  // Log
 
-  std::cout << "Single source : " << source_vect.back() << "\n";
-  print::head(bc_values, 40, "GPU bc values");
-  std::cout << "GPU Elapsed Time : " << run_times[params.num_runs - 1]
-            << " (ms)" << std::endl;
-
-  // --
-  // Run performance evaluation
+#if !(ESSENTIALS_COLLECT_METRICS)
+  // Standard run
+  for (int i = 0; i < n_runs; i++) {
+    run_times.push_back(
+        gunrock::bc::run(G, source_vect[i], bc_values.data().get()));
+  }
+  // Export metrics (runtimes only)
+  if (params.export_metrics) {
+    std::vector<int> empty_vector;
+    gunrock::util::stats::export_performance_stats(
+        empty_vector, empty_vector, n_edges, n_vertices, empty_vector,
+        run_times, "bc", params.filename, "market", params.json_dir,
+        params.json_file, source_vect, tag_vect, num_arguments, argument_array);
+  }
+#else
+  // Run with performance evaluation
+  std::vector<int> edges_visited_vect(n_runs);
+  std::vector<int> search_depth_vect(n_runs);
+  std::vector<int> vertices_visited_vect(n_runs);
 
-  if (params.collect_metrics) {
-    std::vector<int> edges_visited_vect;
-    std::vector<int> search_depth_vect;
-    std::vector<int> nodes_visited_vect(source_vect.size());
+  for (int i = 0; i < n_runs; i++) {
+    benchmark::INIT_BENCH();
 
-    vertex_t n_edges = G.get_number_of_edges();
+    run_times.push_back(
+        gunrock::bc::run(G, source_vect[i], bc_values.data().get()));
 
-    for (int i = 0; i < source_vect.size(); i++) {
-      float metrics_run_time = gunrock::bc::run(
-          G, source_vect[i], params.collect_metrics, bc_values.data().get(),
-          &edges_visited, &search_depth);
+    thrust::host_vector<int> h_edges_visited = benchmark::____.edges_visited;
+    thrust::host_vector<int> h_vertices_visited =
+        benchmark::____.vertices_visited;
 
-      edges_visited_vect.push_back(edges_visited);
-      search_depth_vect.push_back(search_depth);
-    }
+    edges_visited_vect[i] = h_edges_visited[0];
+    vertices_visited_vect[i] = h_vertices_visited[0];
+    search_depth_vect[i] = benchmark::____.search_depth;
```

Comment:
I think this should be transparent as well. If we have INIT_BENCH, DESTROY_BENCH, we can just create another wrapper which moves the data to host: `benchmark::EXTRACT()`? And _that_ returns a well-defined struct that contains the benchmarked metrics. The struct can then be document well to explain what each variable means and how it was obtained (which will be the only interface the user will care about).

- The result of the benchmarks (the struct) can be passed in to the `export_performance_stats` function as well, without needing `empty_vector`, because the struct can have a set default when ESSENTIALS_COLLECT_METRICS was false.
- It will also remove the need for creating std::vectors, etc.